rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+
+This proposal will require the creation of a user settings page _or_ an analytics opt-out checkbox on the `/privacy` page.
+
+### Sample implementaton and event types
+
+Below I've created a TypeScript type for all the proposed anlytics events, and a simple function for sending events. For each event you'll see the payload alongside the default values which will be passed to _all_ events.
+
+```ts
+// src/utils/events.ts
+
+// This type is only for illustrative purposes and would be imported from elsewhere.
+type MediaType = "image" | "audio" | "all";
+
+/**
+ * A list of all events allowed by our analytics server.
+ */
+type AnalyticsEvents = {
+  /** A search performed by the user **/
+  SUBMIT_SEARCH: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** The user navigates to a page. */
+  VIEW_PAGE: {
+    /** The title of the page being navigated to (window.document.title)  */
+    title: string;
+    /** The path of the previous page */
+    previous: string;
+    /** The path of the new page */
+    next: string;
+  };
+  /** Whenever the user sets a filter  */
+  APPLY_FILTER: {
+    /** The name of the filter (not the user-facing element label) */
+    key: string;
+    /** The value of the filter */
+    value: string | boolean;
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** Whenever a user changes the content type */
+  CHANGE_CONTENT_TYPE: {
+    /** The previously-set media type */
+    previous: MediaType;
+    /** The new media type */
+    next: MediaType;
+    /** The name of the Vue component used to switch content types. **/
+    component: string;
+  };
+  /**
+   * Whenever the user scrolls to the end of the results page.
+   * Useful to evaluate how often users load more results or click
+   * on the external sources dropdown.
+   * */
+  REACH_RESULT_END: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** Whenever the user clicks the load more button */
+  LOAD_MORE_RESULTS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** When a user opens the external sources popover. */
+  VIEW_EXTERNAL_SOURCES: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** When the user chooses an external source from the dropdown of external sources */
+  SELECT_EXTERNAL_SOURCE: {
+    /** The name of the external source */
+    name: string;
+    /** The full URL of the source */
+    link: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /**  The user opens the menu which lists pages. */
+  OPEN_PAGES_MENU: {};
```

Comment:
How about also adding events for clicking on external links in the navigation or content pages? Get involved, API, WordPress.org links and the links to the providers on the About page.

Reply:
Great idea, added in 853028f.